src/Base/Exception.h

```diff
@@ -33,6 +33,13 @@
 #include "FileInfo.h"
 #include "BaseClass.h"
 
+/* MACROS FOR THROWING EXCEPTIONS */
+
+#define THROW(exception) throw Base::exception(__FILE__,__LINE__,__PRETTY_FUNCTION__);
+#define THROWM(exception, message) throw Base::exception(message,__FILE__,__LINE__,__PRETTY_FUNCTION__);

```

Comment:
I don't think C++ is forcing you to copy+paste at all.  Another option would be to have the macro translate the N arguments to a single string, so `raise some_exception(FANCY_EXCEPTION_FORMATTER("translated message", __FILE__, __LINE__, etc))`.  Then, the part that needs separate debug/message information can split out the translated message.

Templates could provide another set of solutions.

Or, as Werner suggested, a base class setter, which is used by the macro to set the additional arguments.

But, in the end, I don't really understand the desire to put the method, line, etc in the message in the first place.  On the surface, it seems like a lot of code that will need to be maintained (and, remember all these static strings are taking up executable space - you're effectively adding debug information in to the release build for each usage of THROW()) to replace a quick search for the exception message and/or type.  But, as I mentioned on the forum, I haven't read the whole backstory so could be missing something.